File: packages/core/src/query.ts
// query.ts — the canonical query state shape.
//
// We lean into SQL vocabulary at every layer (per design feedback): a QueryState
// reads top-to-bottom like the statement it compiles to —
//
//     SELECT <select>  WHERE <where>  ORDER BY <orderBy>  LIMIT <limit> OFFSET <offset>
//
// Both "what data" (where/orderBy/limit/offset) and "how viewed" (select +
// per-column width) can be serialized together when a consumer explicitly opts
// into URL synchronization.

/** All filter operators across both source projects, unioned. Every value can
 *  be NULL, so `is_null`/`is_not_null` are valid for every field type. */
export type FilterOp =
  // comparison — number / datetime (and =,!= for any scalar)
  | "="
  | "!="
  | ">"
  | ">="
  | "<"
  | "<="
  // text matching
  | "contains"
  | "starts_with"
  | "ends_with"
  // array (textarray) containment
  | "includes"
  // nullity — valid for EVERY type; for arrays, "empty" vs "non-empty"
  | "is_null"
  | "is_not_null";

/** A single AND-combined filter. `value` is always the raw string the UI
 *  captured; coercion to number/bool/date happens at apply/compile time based on
 *  the field's declared type (never on the value's runtime shape). Unused for
 *  the nullary ops (`is_null`/`is_not_null`). */
export interface WhereClause {
  field: string; // FieldDef.name
  op: FilterOp;
  value: string;
}

/** One ORDER BY term. Array order in `QueryState.orderBy` is the sort priority. */
export interface OrderByClause {
  field: string; // FieldDef.name (server may sort a different expr via FieldDef.sort.field)
  dir: "asc" | "desc";
  /** NULL placement. Omitted = "last" (the historical default both projects used,
   *  preserved so old `?q=` links round-trip identically). */
  nulls?: "first" | "last";
}

/** One column in the SELECT list, plus the view state that travels with it.
 *  Order in `QueryState.select` is the left-to-right display order. */
export interface SelectColumn {
  field: string; // FieldDef.name
  /** Pixel width override. Omitted = FieldDef.select.width, else the type default. */
  width?: number;
}

/** A single aggregate operation. `count` is the only op that needs no measure
 *  column (it counts rows); the rest reduce one column's values. The server-side
 *  matrix (backends/go aggOpAllowed + the TS AGG_OPS_BY_TYPE) decides which ops a
 *  field's type permits. */
export type AggOp = "count" | "count_distinct" | "sum" | "avg" | "min" | "max";

/** One optional "metric" pinned above the table: a single aggregate over a single
 *  measure column, optionally broken down by one or more group columns.
 *
 *  Scope is deliberately the *whole filtered set* — the same WHERE as the table,
 *  but NOT its ORDER BY / LIMIT / OFFSET. It is always evaluated on the server
 *  (a real GROUP BY), so it reflects every matching row, not just the visible
 *  page. Lives inside QueryState so it round-trips through `?q=`, saved queries,
 *  and undo/redo for free — a saved query is a saved dashboard. */
export interface AggregationClause {
  /** Stable id; keys the metric panel and survives a `?q=` round-trip. */
  id: string;
  op: AggOp;
  /** Measure column (FieldDef.name). Omit only for `count` (⇒ COUNT(*)). */
  field?: string;
  /** Group-by columns, in axis order. `[]` = a single grand-total value.
   *  One ⇒ a list/bar breakdown; two ⇒ an x/y pivot; three+ ⇒ a flat table. */
  groupBy: string[];
  /** Panel header override; defaults to a derived label like "avg total". */
  label?: string;
}

export interface QueryState {
  /** Ordered SELECT list + per-column widths. Empty = the schema's default columns. */
  select: SelectColumn[];
  /** AND-combined filters. */
  where: WhereClause[];
  /** Multi-sort terms in priority order. Empty = the schema's default sort. */
  orderBy: OrderByClause[];
  /** Page size. */
  limit: number;
  /** Page offset (rows). */
  offset: number;
  /** Optional aggregate metrics shown above the table. Omitted/empty = none.
   *  Each runs as its own server GROUP BY over the WHERE-filtered set (no paging). */
  aggregations?: AggregationClause[];
}

/** A row's stable identity, used for selection and per-row refresh. */
export type RowId = string | number;

/** The query a fresh table starts from before schema defaults are layered on.
 *  Empty `select`/`orderBy`/`where` mean "defer to the schema", which also keeps
 *  encoded `?q=` tokens short (defaults are omitted). */
export const EMPTY_QUERY: QueryState = {
  select: [],
  where: [],
  orderBy: [],
  limit: 100,
  offset: 0,
};

/** Structural limits applied whenever query state crosses a trust boundary
 * (URL, storage, an imperative setQuery call, or a server projection). Row
 * limits may use the full safe-integer range; consumers and their backends own
 * any smaller operational cap appropriate for their dataset. */
export const MAX_QUERY_LIMIT = Number.MAX_SAFE_INTEGER;
export const MAX_QUERY_OFFSET = 1_000_000;
export const MAX_SELECT_COLUMNS = 200;
export const MAX_WHERE_CLAUSES = 100;
export const MAX_ORDER_BY_TERMS = 20;
export const MAX_AGGREGATIONS = 20;
export const MAX_GROUP_BY_FIELDS = 20;
export const MAX_QUERY_TOKEN_LENGTH = 2 * 1024 * 1024;

const MAX_FIELD_NAME_LENGTH = 256;
const MAX_FILTER_VALUE_LENGTH = 10_000;
const MAX_LABEL_LENGTH = 1_000;
const MIN_COLUMN_WIDTH = 24;
const MAX_COLUMN_WIDTH = 2_000;

const FILTER_OPS: ReadonlySet<string> = new Set([
  "=",
  "!=",
  ">",
  ">=",
  "<",
  "<=",
  "contains",
  "starts_with",
  "ends_with",
  "includes",
  "is_null",
  "is_not_null",
]);

const AGG_OPS: ReadonlySet<string> = new Set(["count", "count_distinct", "sum", "avg", "min", "max"]);

function isRecord(value: unknown): value is Record<string, unknown> {
  return value != null && typeof value === "object" && !Array.isArray(value);
}

function boundedString(value: unknown, maxLength: number): string | null {
  return typeof value === "string" && value.length > 0 && value.length <= maxLength ? value : null;
}

function boundedInteger(value: unknown, fallback: number, min: number, max: number): number {
  if (typeof value !== "number" || !Number.isFinite(value)) return fallback;
  return Math.max(min, Math.min(max, Math.round(value)));
}

/** Convert unknown input into a bounded, structurally valid QueryState.
 * Invalid members are dropped; invalid scalar paging values use the supplied
 * fallback. This function is intentionally schema-agnostic—field allowlisting
 * still happens in toServerQuery and in the backend compiler. */
export function normalizeQueryState(input: unknown, fallback: QueryState = EMPTY_QUERY): QueryState {
  const raw = isRecord(input) ? input : {};

  const select: SelectColumn[] = [];
  if (Array.isArray(raw.select)) {
    for (const item of raw.select.slice(0, MAX_SELECT_COLUMNS)) {
      if (!isRecord(item)) continue;
      const field = boundedString(item.field, MAX_FIELD_NAME_LENGTH);
      if (!field) continue;
      const column: SelectColumn = { field };
      if (typeof item.width === "number" && Number.isFinite(item.width)) {
        column.width = boundedInteger(item.width, MIN_COLUMN_WIDTH, MIN_COLUMN_WIDTH, MAX_COLUMN_WIDTH);
      }
      select.push(column);
    }
  } else {
    select.push(...fallback.select.map((column) => ({ ...column })));
  }

  const where: WhereClause[] = [];
  if (Array.isArray(raw.where)) {
    for (const item of raw.where.slice(0, MAX_WHERE_CLAUSES)) {
      if (!isRecord(item)) continue;
      const field = boundedString(item.field, MAX_FIELD_NAME_LENGTH);
      const value = typeof item.value === "string" && item.value.length <= MAX_FILTER_VALUE_LENGTH ? item.value : null;
      if (!field || typeof item.op !== "string" || !FILTER_OPS.has(item.op) || value == null) continue;
      where.push({ field, op: item.op as FilterOp, value });
    }
  } else {
    where.push(...fallback.where.map((clause) => ({ ...clause })));
  }

  const orderBy: OrderByClause[] = [];
  if (Array.isArray(raw.orderBy)) {
    for (const item of raw.orderBy.slice(0, MAX_ORDER_BY_TERMS)) {
      if (!isRecord(item)) continue;
      const field = boundedString(item.field, MAX_FIELD_NAME_LENGTH);
      if (!field || (item.dir !== "asc" && item.dir !== "desc")) continue;
      const term: OrderByClause = { field, dir: item.dir };
      if (item.nulls === "first" || item.nulls === "last") term.nulls = item.nulls;
      orderBy.push(term);
    }
  } else {
    orderBy.push(...fallback.orderBy.map((term) => ({ ...term })));
  }

  const out: QueryState = {
    select,
    where,
    orderBy,
    limit: boundedInteger(raw.limit, fallback.limit, 1, MAX_QUERY_LIMIT),
    offset: boundedInteger(raw.offset, fallback.offset, 0, MAX_QUERY_OFFSET),
  };

  if (Array.isArray(raw.aggregations)) {
    const aggregations: AggregationClause[] = [];
    for (const item of raw.aggregations.slice(0, MAX_AGGREGATIONS)) {
      if (!isRecord(item)) continue;
      const id = boundedString(item.id, MAX_FIELD_NAME_LENGTH);
      if (!id || typeof item.op !== "string" || !AGG_OPS.has(item.op) || !Array.isArray(item.groupBy)) continue;
      const groupBy = item.groupBy
        .slice(0, MAX_GROUP_BY_FIELDS)
        .map((field) => boundedString(field, MAX_FIELD_NAME_LENGTH))
        .filter((field): field is string => field != null);
      const aggregation: AggregationClause = { id, op: item.op as AggOp, groupBy };
      const field = boundedString(item.field, MAX_FIELD_NAME_LENGTH);
      const label = boundedString(item.label, MAX_LABEL_LENGTH);
      if (field) aggregation.field = field;
      if (label) aggregation.label = label;
      aggregations.push(aggregation);
    }
    if (aggregations.length > 0) out.aggregations = aggregations;
  } else if (fallback.aggregations?.length) {
    out.aggregations = fallback.aggregations.map((aggregation) => ({ ...aggregation, groupBy: [...aggregation.groupBy] }));
  }

  return out;
}

/** Structural equality on two queries — used to decide whether a `?q=` write or a
 *  refetch is actually needed (avoids history spam / redundant fetches). */
export function queriesEqual(a: QueryState, b: QueryState): boolean {
  if (a === b) return true;
  if (a.limit !== b.limit || a.offset !== b.offset) return false;
  if (!sameArray(a.where, b.where)) return false;
  if (!sameArray(a.orderBy, b.orderBy)) return false;
  if (!sameArray(a.select, b.select)) return false;
  if (!sameArray(a.aggregations ?? [], b.aggregations ?? [])) return false;
  return true;
}

function sameArray<T>(a: readonly T[], b: readonly T[]): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    // Members are flat records of primitives — JSON compare is correct and cheap here.
    if (JSON.stringify(a[i]) !== JSON.stringify(b[i])) return false;
  }
  return true;
}

File: backends/go/query.go
// Package querytable compiles a frontend QueryState into parameterized SQL,
// validated against a server-defined field schema. It is the Go projection of
// @pythia-software/query-table-core: the wire types here mirror the TS shapes exactly so a
// base64url ?q= token (or a JSON request body) round-trips without translation.
//
// Safety model: the column
// expression for each field comes from the schema, never from request input. An
// unknown field is a compile error before any SQL is built. Values reach SQL
// only as bound placeholders ($N). The validated operator and the schema-defined
// expression are the only request-influenced tokens in the SQL text.
package querytable

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
)

// Resource limits mirror @pythia-software/query-table-core. Compile and DecodeWireQuery both
// enforce them so callers are protected whether a query came from a URL token
// or was decoded from a JSON request body elsewhere.
const (
	MaxQueryLimit       = 1_000
	MaxQueryOffset      = 1_000_000
	MaxSelectColumns    = 200
	MaxWhereClauses     = 100
	MaxOrderByTerms     = 20
	MaxAggregations     = 20
	MaxGroupByFields    = 20
	maxFieldNameLength  = 256
	maxFilterValueBytes = 10_000
	maxQueryTokenBytes  = 2 * 1024 * 1024
)

// WhereClause mirrors @pythia-software/query-table-core WhereClause.
type WhereClause struct {
	Field string `json:"field"`
	Op    string `json:"op"`
	Value string `json:"value"`
}

// OrderBy mirrors @pythia-software/query-table-core OrderByClause.
type OrderBy struct {
	Field string `json:"field"`
	Dir   string `json:"dir"`             // "asc" | "desc"
	Nulls string `json:"nulls,omitempty"` // "first" | "last" | "" (default last)
}

// AggSpec mirrors @pythia-software/query-table-core AggregationClause: one aggregate op over one
// measure column (Field; empty ⇒ COUNT(*)), broken down by zero or more group
// columns. Compiled by CompileAggregation; the metric panel runs one per spec
// over the WHERE-filtered set (no paging).
type AggSpec struct {
	ID      string   `json:"id"`
	Op      string   `json:"op"`
	Field   string   `json:"field,omitempty"`
	GroupBy []string `json:"groupBy,omitempty"`
}

// WireQuery is the server-bound subset of QueryState (the TS ServerQuery / the
// {s,w,o,...} `?q=` payload). View-only state (column widths) never arrives.
//
// OrderBy unmarshals from BOTH the new array form and the legacy single-object
// form, so legacy links keep compiling.
type WireQuery struct {
	Select       []string      `json:"select,omitempty"`
	Where        []WhereClause `json:"w,omitempty"`
	OrderBy      OrderBys      `json:"o,omitempty"`
	Limit        int           `json:"l,omitempty"`
	Offset       int           `json:"f,omitempty"`
	Aggregations []AggSpec     `json:"g,omitempty"`
}

// UnmarshalJSON accepts both the readable ServerQuery property names emitted
// by @pythia-software/query-table-core and the compact names used inside a URL token. It
// validates resource limits before returning, so a normal json.Decoder is a
// safe API boundary even when the caller does not use DecodeWireQuery.
func (q *WireQuery) UnmarshalJSON(data []byte) error {
	var payload struct {
		Select              []string      `json:"select"`
		Where               []WhereClause `json:"where"`
		CompactWhere        []WhereClause `json:"w"`
		OrderBy             OrderBys      `json:"orderBy"`
		CompactOrderBy      OrderBys      `json:"o"`
		Limit               *int          `json:"limit"`
		CompactLimit        *int          `json:"l"`
		Offset              *int          `json:"offset"`
		CompactOffset       *int          `json:"f"`
		Aggregations        []AggSpec     `json:"aggregations"`
		CompactAggregations []AggSpec     `json:"g"`
	}
	if err := json.Unmarshal(data, &payload); err != nil {
		return err
	}

	q.Select = payload.Select
	q.Where = payload.Where
	if q.Where == nil {
		q.Where = payload.CompactWhere
	}
	q.OrderBy = payload.OrderBy
	if q.OrderBy == nil {
		q.OrderBy = payload.CompactOrderBy
	}
	q.Limit = 0
	if payload.Limit != nil {
		q.Limit = *payload.Limit
	} else if payload.CompactLimit != nil {
		q.Limit = *payload.CompactLimit
	}
	q.Offset = 0
	if payload.Offset != nil {
		q.Offset = *payload.Offset
	} else if payload.CompactOffset != nil {
		q.Offset = *payload.CompactOffset
	}
	q.Aggregations = payload.Aggregations
	if q.Aggregations == nil {
		q.Aggregations = payload.CompactAggregations
	}

	return q.Validate()
}

// Validate rejects malformed or unexpectedly expensive query shapes. A zero
// Limit is allowed so an omitted value can be replaced by the caller's default;
// an explicitly negative or oversized value is never allowed.
func (q WireQuery) Validate() error {
	if q.Limit < 0 || q.Limit > MaxQueryLimit {
		return fmt.Errorf("limit must be between 0 and %d", MaxQueryLimit)
	}
	if q.Offset < 0 || q.Offset > MaxQueryOffset {
		return fmt.Errorf("offset must be between 0 and %d", MaxQueryOffset)
	}
	if len(q.Select) > MaxSelectColumns {
		return fmt.Errorf("select has %d fields; maximum is %d", len(q.Select), MaxSelectColumns)
	}
	if len(q.Where) > MaxWhereClauses {
		return fmt.Errorf("where has %d clauses; maximum is %d", len(q.Where), MaxWhereClauses)
	}
	if len(q.OrderBy) > MaxOrderByTerms {
		return fmt.Errorf("orderBy has %d terms; maximum is %d", len(q.OrderBy), MaxOrderByTerms)
	}
	if len(q.Aggregations) > MaxAggregations {
		return fmt.Errorf("aggregations has %d items; maximum is %d", len(q.Aggregations), MaxAggregations)
	}
	for _, field := range q.Select {
		if field == "" || len(field) > maxFieldNameLength {
			return fmt.Errorf("invalid select field name")
		}
	}
	for _, clause := range q.Where {
		if clause.Field == "" || len(clause.Field) > maxFieldNameLength {
			return fmt.Errorf("invalid filter field name")
		}
		if len(clause.Value) > maxFilterValueBytes {
			return fmt.Errorf("filter value for %q exceeds %d bytes", clause.Field, maxFilterValueBytes)
		}
	}
	for _, term := range q.OrderBy {
		if term.Field == "" || len(term.Field) > maxFieldNameLength {
			return fmt.Errorf("invalid sort field name")
		}
	}
	for _, aggregation := range q.Aggregations {
		if aggregation.ID == "" || len(aggregation.ID) > maxFieldNameLength {
			return fmt.Errorf("invalid aggregation id")
		}
		if len(aggregation.GroupBy) > MaxGroupByFields {
			return fmt.Errorf("aggregation %q has %d group fields; maximum is %d", aggregation.ID, len(aggregation.GroupBy), MaxGroupByFields)
		}
	}
	return nil
}

// OrderBys is a slice of OrderBy that also accepts a single object on decode
// (legacy compatibility) and a base64url `c`/`s` is handled at the WireQuery
// level via DecodeWireQuery.
type OrderBys []OrderBy

// UnmarshalJSON accepts either `[{...},{...}]` (current) or `{...}` (legacy).
func (o *OrderBys) UnmarshalJSON(b []byte) error {
	trimmed := strings.TrimSpace(string(b))
	if trimmed == "null" || trimmed == "" {
		*o = nil
		return nil
	}
	if strings.HasPrefix(trimmed, "[") {
		var arr []OrderBy
		if err := json.Unmarshal(b, &arr); err != nil {
			return err
		}
		*o = arr
		return nil
	}
	var single OrderBy
	if err := json.Unmarshal(b, &single); err != nil {
		return err
	}
	*o = []OrderBy{single}
	return nil
}

// wirePayload is the compact JSON the frontend base64url-encodes. `c` is the
// legacy alias for select column names; `s` is the current select (tuples).
type wirePayload struct {
	Select json.RawMessage `json:"s,omitempty"` // [field] | [field,width] tuples, or legacy string[]
	Legacy json.RawMessage `json:"c,omitempty"` // legacy string[]
	Where  []WhereClause   `json:"w,omitempty"`
	Order  OrderBys        `json:"o,omitempty"`
	Limit  int             `json:"l,omitempty"`
	Offset int             `json:"f,omitempty"`
	Aggs   []AggSpec       `json:"g,omitempty"`
}

// DecodeWireQuery decodes the base64url-encoded JSON `?q=` token produced by
// @pythia-software/query-table-core encodeQuery. The select tuples carry widths the server
// ignores; only the field names are extracted. Mirrors the charset/padding
// fix-ups so a bookmark round-trips bit-for-bit.
func DecodeWireQuery(token string) (WireQuery, error) {
	var q WireQuery
	if token == "" {
		return q, nil
	}
	if len(token) > maxQueryTokenBytes {
		return q, fmt.Errorf("query token exceeds %d bytes", maxQueryTokenBytes)
	}
	s := strings.ReplaceAll(token, "-", "+")
	s = strings.ReplaceAll(s, "_", "/")
	if pad := len(s) % 4; pad != 0 {
		s += strings.Repeat("=", 4-pad)
	}
	raw, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		return q, fmt.Errorf("base64: %w", err)
	}
	var p wirePayload
	if err := json.Unmarshal(raw, &p); err != nil {
		return q, fmt.Errorf("json: %w", err)
	}
	q.Where = p.Where
	q.OrderBy = p.Order
	q.Limit = p.Limit
	q.Offset = p.Offset
	q.Aggregations = p.Aggs
	q.Select = decodeSelect(p.Select, p.Legacy)
	if err := q.Validate(); err != nil {
		return WireQuery{}, fmt.Errorf("query: %w", err)
	}
	return q, nil
}

// decodeSelect pulls field names from the `s` tuples (or legacy `c`/`s` string[]).
func decodeSelect(s, legacy json.RawMessage) []string {
	for _, raw := range []json.RawMessage{s, legacy} {
		if len(raw) == 0 {
			continue
		}
		// Try string[] first (legacy), then tuple form.
		var names []string
		if err := json.Unmarshal(raw, &names); err == nil {
			return names
		}
		var tuples [][]json.RawMessage
		if err := json.Unmarshal(raw, &tuples); err == nil {
			out := make([]string, 0, len(tuples))
			for _, t := range tuples {
				if len(t) == 0 {
					continue
				}
				var name string
				if err := json.Unmarshal(t[0], &name); err == nil {
					out = append(out, name)
				}
			}
			return out
		}
	}
	return nil
}
